### ISSUE TITLE
Vertical model coordinate with more than one dimension

### ISSUE THREAD

**Issue by wachsylon** [PCMDI/cmor#499]
Hi,
I have data for the variable `rld` in `CFsubhr`. This uses dimensions `alevhalf` and `site`.
Is it possible that the vertical model coordinate depends also on `site`? How can I configure CMOR for that? 
Best regards,
Fabi

**Comment by taylor13** [comment]
I think you have identified a issue here that we should try to solve.  If you want to report output on model levels (using, for example, hybrid sigma coordinate), then you need to include surface pressure (ps) as a "formula term" (aka a "zfactor").  In this case ps will vary as a function of "site", but we don't have an entry in CMIP6_formula_terms.json with ps(site, level, time), so we can't write this variable.  We *might* be able to fix this by just adding such an entry to the formula_terms table.

I note that ps(site,time1) already is available in CFsubhr, so we could copy that into CMIP6_formula_terms.json, giving it a different "entry" name (e.g., "pssite"), but with outname "ps".  Do you think that would work?

**Comment by taylor13** [comment]
Further clarification:  The actual parametric vertical coordinate must have the same values for all sites; only certain parameter terms (like ps) can vary with site.  

Also, I now recall that this case should be handled using the "cmor_grid" function.  The axis_id passed to cmor_grid would point to the "site" axis. I think 3-d fields  written this way still will need to access in the formula terms table a "pssite" variable, as discussed above. 

@mauzey1 @wachsyon: could one of you devise a test case to see that this works?

thanks,
Karl

**Comment by taylor13** [comment]
I just found a test code that should already tell us that this works: https://github.com/PCMDI/cmor/blob/master/Test/test_python_cfmip_site_axis_test.py  .  This is for vertical coordinate that needs orog(site), rather than ps(site,time1), but I think it is probably a sufficient test.  The only thing I don't know is whether "orog" appears in the test table as "orog(site)" or "orog(lon, lat)" .  With the grid definition approach, I don't think we need "orog(site)"  (or "pssite(site,time)").

**Comment by mauzey1** [comment]
`orog` has dimensions `longitude` and `latitude` in CMIP6_formula_terms.json
```
        "orog": {
            "long_name": "Surface Altitude", 
            "units": "m", 
            "dimensions": "longitude latitude", 
            "out_name": "orog", 
            "type": "real"
        }
```

But in the output of test_python_cfmip_site_axis_test.py, it has the dimension `site`
```
        float orog(site) ;
                orog:long_name = "Surface Altitude" ;
                orog:units = "m" ;
```

**Comment by wachsylon** [comment]
>  The actual parametric vertical coordinate must have the same values for all sites

Is this a general necessary condition? This was my main question. I thought that, if `site`s represent stations,  e.g. pressure levels may vary from `site` to `site`.

This also overlaps with #494 where the model's SLEVE coordinate is "corrected" with additonal terms. Due to that, the `z_factor`s are not sufficient to fully describe the vertical coordinate. Why can we not allow to provide a 3d coordinate, in that specific case `height(lon,lat,lev)`?

**Comment by wachsylon** [comment]
This is also related to the request of Robert Pincus from University of Colorado/NOAA Earth System Research Lab. In his data from a non-GCM radiative transfer model, pressure level vary from site to site.

**Comment by taylor13** [comment]
I think the kludge we anticipated to handle this for variables in CFsubhr was to not provide all the information in a single file.  Rather we would write "rld" in one file and then "phalf" (or "pfull"?) in another.  These variables would be a function of time, site, and alevel.  In this case alevel could be either an "index" or a "pseudo coordinate containing some nominal pressure values (if all the sites have pressure levels that are about the same). 

I think to implement such a solution, we probably need to add to the coordinates.json file a new vertical coordinate (e.g., "level_index" or "nominal_pressure"). 

Do you think this would work?

**Comment by wachsylon** [comment]
But if a part of a vertical coordinate is missing in the file, the file is no longer self-descriptive. On the other hand, I just found  CMIP6.CMIP.NOAA-GFDL.GFDL-AM4.amip.r1i1p1f1.Amon.cli.gr1  in ESGF with
```
	double lev(lev) ;
		lev:standard_name = "atmosphere_hybrid_sigma_pressure_coordinate" ;
...
// global attributes:
		:external_variables = "ps" ;
```
where `ps` is not saved within the file. PrePARE definitely gives no errors for that. This makes me a bit tired because we spent about months on how to add `ps` to all files of variables that use the `hybrid_sigma_pressure` axis and process those with CMOR.

So is it O.K. to save a `z_factor` or `plev`s in an additional file? It would do save storage space and post-processing time...

> Also, I now recall that this case should be handled using the "cmor_grid" function

I will do some tests but unfortunately I cannot start them until next week.
